R/TunerIrace.R

```diff
@@ -0,0 +1,148 @@
+#'   task = tsk("iris"),
+#'   learner = lrn("classif.rpart"),
+#'   resampling = rsmp("holdout"),
+#'   measure = msr("classif.ce"),
+#'   search_space = search_space,
+#'   terminator = terminator
+#' )
+#' tt = tnr("irace")
+#'
+#' # modifies the instance by reference
+#' tt$optimize(instance)
+#'
+#' # returns best configuration and best performance
+#' instance$result
+#'
+#' # allows access of data.table of full path of all evaluations
+#' instance$archive
+TunerIrace = R6Class("TunerIrace",
+  inherit = Tuner,
+  public = list(
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function() {
+      ps = ParamSet$new(list(
+        ParamInt$new("n_instances", lower = 1, default = 10),
+        ParamInt$new("debugLevel", default = 0, lower = 0),
+        ParamInt$new("seed"),
+        ParamDbl$new("postselection", default = 0, lower = 0, upper = 1),
+        ParamInt$new("elitist", default = 1, lower = 0, upper = 1),
+        ParamInt$new("elitistLimit", default = 2, lower = 0),
+        ParamInt$new("nbIterations", default = 0, lower = 0),
+        ParamInt$new("nbExperimentsPerIteration", default = 0, lower = 0),
+        ParamInt$new("minNbSurvival", default = 0, lower = 0),
+        ParamInt$new("nbConfigurations", default = 0, lower = 0),
+        ParamInt$new("mu", default = 5, lower = 1),
+        ParamInt$new("softRestart", default = 1, lower = 0, upper = 1),
+        ParamDbl$new("softRestartThreshold"),
+        ParamInt$new("digits", default = 4, lower = 1, upper = 15),
+        ParamFct$new("testType", default = "F-test",
+          levels = c("F-test", "t-test", "t-test-bonferroni", "t-test-holm")),
+        ParamInt$new("firstTest", default = 5, lower = 0),
+        ParamInt$new("eachTest", default = 1, lower = 1),
+        ParamDbl$new("confidence", default = 0.95, lower = 0, upper = 1),
+        ParamInt$new("capping", default = 0, lower = 0, upper = 1),
+        ParamFct$new("cappingType", default = "median", levels = c("median", "mean", "best", "worst")),
+        ParamFct$new("boundType", default = "candidate", levels = c("candidate", "instance")),
+        ParamDbl$new("boundMax", default = 0),
+        ParamInt$new("boundDigits", default = 0),
+        ParamDbl$new("boundPar", default = 1),
+        ParamDbl$new("boundAsTimeout", default = 1)
+      ))
+      ps$values = list(n_instances = 10)
+
+      super$initialize(
+        param_set = ps,
+        param_classes = c("ParamDbl", "ParamInt", "ParamFct", "ParamLgl"),
+        properties = c("dependencies", "single-crit"),
+        packages = "irace"
+      )
+    }
+  ),
+
+  private = list(
+    .optimize = function(inst) {
+      pv = self$param_set$values
+      terminator = inst$terminator
+      objective = inst$objective
+      
+      # Set resampling instances
+      ri = map(seq(pv$n_instances), function(n) {
+        r = objective$resampling$clone()
+        r$instantiate(objective$task)
```

Comment:
As defined in the objective the resampling is already instantiated. Why doing it here again? Do we want the same resampling splits in all scenarios for paired testing or do we specifically want the opposite? Anyhow, those things should be commented!

Reply:
We want to give `irace` n different resampling train/ test splits. `irace` then decides which configurations are evaluated on which resamplings. We could also pass `1:n` here and the number is the seed of the train/ test split. But I think this is more complicated.

Reply:
See `target_runner()`. `irace` gives us the  configuration `experiment$configuration` and the resampling `experiment$instance` on which have to evaluate the configuration.